jena-fuseki2/jena-fuseki-docker/README.md

```diff
@@ -0,0 +1,96 @@
+# Apache Jena Fuseki Docker Tools
+
+This package contains a Dockerfile, docker-compose file, and helper scripts to
+create a docker container for Apache Jena Fuseki.
+
+The docker container is based on 
+[Fuseki main](https://jena.apache.org/documentation/fuseki2/fuseki-main)
+for running a SPARQL server.
+
+There is no UI - all configuration is by command line and all usage by via the
+network protocols.
+
+Databases can be mounted outside the docker container so they are preserved when
+the container terminates.
+
+This build system allows the user to customize the docker image.
+
+The docker build downloads the server binary from 
+[Maven central](https://repo1.maven.org/maven2/org/apache/jena/jena-fuseki-server/),
+checking the download against the SHA1 checksum.
+
+## Database
+
+There is a volume mapping "./databases" in the current diectory into the server.
+This can be used to contain databases outside, but accessible to, the container
+that do not get deleted when the container exits.
+
+See examples below.
+
+## Build
+
+Choose the version number of Apache Jena release you wish to use. This toolkit
+defaults to the version of the overall Jena release it was part of. It is best
+to use the release of this set of tools from the same release of the desired
+server.
+
+    docker-compose build --build-arg JENA_VERSION=3.16.0
+
+Note the build command must provide the version number.
+
+## Run
+
+Examples:
+
+Start Fuseki with an in-memory, updatable dataset at http://<i>host</i>:3030/ds
+
+    docker-compose run --rm --service-ports fuseki --mem /ds
+
+Load a TDB2 database, and expose, read-only, via docker:
+
+    mkdir -p databases/DB2
+    tdb2.tdbloader --loc databases/DB2 MyData.ttl
+    # Publish read-only
+    docker-compose run --rm --name MyServer --service-ports fuseki --tdb2 --loc databases/DB2 /ds
```

Comment:
Why not `docker-compose up`?

Reply:
To pass in arguments without modifying the `docker-compose.yaml` and also in the same style as running Fuseki not in docker.

Reply:
But the [Compose FAQ says](https://docs.docker.com/compose/faq/#whats-the-difference-between-up-run-and-start):
>Typically, you want `docker-compose up`. Use up to start or restart all the services defined in a `docker-compose.yml`. In the default “attached” mode, you see all the logs from all the containers. In “detached” mode (`-d`), Compose exits after starting the containers, but the containers continue to run in the background.
> The `docker-compose run` command is for running “one-off” or “adhoc” tasks. It requires the service name you want to run and only starts containers for services that the running service depends on. Use `run` to run tests or perform an administrative task such as removing or adding data to a data volume container.

Sounds to me like running a database is *not* a one-off or administrative task, as you normally want to persist the data, and therefore `docker-compose up` is a better fit.

Could we at least have alternative examples with `up` and `run`?

Reply:
You can't pass arguments to a specific container with `up` except by rewriting to use environment variables, at least as far as I know. The simple way to use the container is to use `run` with the standard Fuseki argument line.

If you wish to contribute, say, a tutorial on using docker-compose, that would be excellent.

JENA-1949 focuses on having a build kit in the codebase and does not try to do everything. Teh description includes "This ticket is for the setup and scripts to build an image, not to publish it." I believe it is a step forward.

Reply:
I think the `up` version would be as simple as copying the same arguments into `command` in `docker-compose.yml`:

    fuseki:
        command: [ "--rm", "--name", "MyServer", "--service-ports", "fuseki", "--tdb2", "--loc", "databases/DB2", "/ds" ]

https://docs.docker.com/compose/compose-file/compose-file-v2/#command

I'm using `up` all the time because I just want to launch all the (say 4-5) interdependent services that comprise the project, at the same time. And I definitely prefer having the arguments in a config file (`docker-compose.yml`) rather than typing or pasting them into the shell, as one would have to using `run`.